kotlinx.fuzz.api/src/main/kotlin/kotlinx/fuzz/KFuzzConfig.kt

```diff
@@ -1,31 +1,145 @@
+        "kotlinx.fuzz.customHookExcludes",
+        defaultValue = emptyList(),
+        toString = { it.joinToString(",") },
+        fromString = { it.split(",") },
+    )
+    override var maxSingleTargetFuzzTime: Duration by KFuzzConfigProperty(
+        "kotlinx.fuzz.maxSingleTargetFuzzTime",
+        toString = { it.inWholeSeconds.toString() },
+        fromString = { it.toInt().seconds },
+    )
+
+    override fun toPropertiesMap(): Map<String, String> = configProperties()
+        .associate { it.systemProperty to it.stringValue }
+
+    private fun assertAllSet() {
+        configProperties().forEach { it.assertIsSet() }
+    }
+
+    private fun getAll() {
+        configProperties().forEach { it.get() }
+    }
+
+    private fun validate() {
+        require(maxSingleTargetFuzzTime.inWholeSeconds > 0) { "'maxSingleTargetFuzzTime' must be at least 1 second" }
+        require(keepGoing > 0) { "'keepGoing' must be positive" }
+    }
+
     companion object {
-        fun fromSystemProperties(): KFuzzConfig = KFuzzConfig(
-            fuzzEngine = System.getProperty("kotlinx.fuzz.engine", "jazzer"),
-            hooks = System.getProperty("kotlinx.fuzz.hooks").toBooleanOrFalse(),
-            instrument = System.getProperty("kotlinx.fuzz.instrument")?.asList().orEmpty(),
-            customHookExcludes = System.getProperty("kotlinx.fuzz.customHookExcludes")?.asList().orEmpty(),
-            maxSingleTargetFuzzTime = System.getProperty("kotlinx.fuzz.maxSingleTargetFuzzTime", "10").toInt().seconds,
-        )
+        fun build(block: KFuzzConfigImpl.() -> Unit): KFuzzConfig = KFuzzConfigImpl().apply {
+            block()
+            assertAllSet()
+            validate()
+        }
+
+        internal fun fromSystemProperties(): KFuzzConfig = KFuzzConfigImpl().apply {
+            getAll()
+            validate()
+        }
+    }
+}
+
+internal class KFuzzConfigProperty<T : Any>(
+    val systemProperty: String,
+    val defaultValue: T? = null,
+    val toString: (T) -> String,
+    val fromString: (String) -> T,
+) : ReadWriteProperty<Any, T> {
+    private var cachedValue: T? = null
+    val stringValue: String get() = toString(get())
+    override fun getValue(thisRef: Any, property: KProperty<*>): T = get()
+
+    internal fun get(): T {
+        cachedValue?.let {
+            return cachedValue!!
+        }
+
+        cachedValue = System.getProperty(systemProperty)?.let(fromString) ?: defaultValue
+            ?: error("No value for property '$systemProperty'")
+
+        return cachedValue!!
+    }
+
+    internal fun assertIsSet() {
+        require(cachedValue != null || defaultValue != null) { "property '$systemProperty' is not set" }
+    }
+
+    override fun setValue(thisRef: Any, property: KProperty<*>, value: T) {
+        cachedValue?.let {
+            error("Property '${property.name}' is already set")
+        }
+        cachedValue = value
     }
```

Comment:
To be honest, the logic seems unnecessarily complicated to me. When we read a property, we will have lazy init from system properties, but when we write a property, we won't write them into system, so we need to do an explicit dump of our config, but not an explicit load. But wait, we have `fromSystemProperties()`, so we actually _do_ need to do an explicit load? And I don't see a method for explicit write to system properties, which I just said we should do?... I'm confused 🤔

Reply:
Yeah, it's complicated. More on it [here](https://github.com/JetBrains-Research/kotlinx.fuzz/pull/23#issuecomment-2607077863)